File: src/visual-graph/prune/iterate-visual-nodes.test.ts
import { describe, expect, test } from "vitest";

import { DIRECTION } from "../direction.js";
import type { Direction } from "../direction.js";
import type { NodeKind } from "../node-kind.js";
import { NODE_KIND } from "../node-kind.js";
import { SUBGRAPH_KIND } from "../subgraph-kind.js";
import { VISUAL_ELEMENT_TYPE } from "../visual-element-type.js";
import type { VisualElement } from "../visual-element.js";
import type { VisualNode } from "../visual-node.js";
import type { VisualSubgraph } from "../visual-subgraph.js";
import { iterateVisualNodes } from "./iterate-visual-nodes.js";

const node = (
  id: string,
  kind: NodeKind = NODE_KIND.ConstBinding,
): VisualNode => {
  const common = {
    type: VISUAL_ELEMENT_TYPE.Node,
    id,
    name: id,
    line: 1,
    isJsxElement: false,
    endLine: null,
    unused: false,
  } as const;
  if (
    kind === NODE_KIND.ConstBinding ||
    kind === NODE_KIND.LetBinding ||
    kind === NODE_KIND.VarBinding
  ) {
    return { ...common, kind, initIsFunction: false };
  }
  if (kind === NODE_KIND.WriteReference) {
    return { ...common, kind, declarationKind: null };
  }
  if (
    kind === NODE_KIND.NamedImportBinding ||
    kind === NODE_KIND.DefaultImportBinding ||
    kind === NODE_KIND.NamespaceImportBinding
  ) {
    throw new Error(
      "ImportBinding fixture not supported by iterate-visual-nodes test",
    );
  }
  return { ...common, kind };
};

const sg = (
  id: string,
  elements: VisualElement[],
  direction: Direction = DIRECTION.TB,
): VisualSubgraph => ({
  type: VISUAL_ELEMENT_TYPE.Subgraph,
  id,
  kind: SUBGRAPH_KIND.Function,
  line: 1,
  direction,
  elements,
  endLine: null,
  ownerNodeId: "n_owner",
  ownerName: "owner",
});

describe("iterateVisualNodes", () => {
  test("yields only ROOT_CANDIDATE_KINDS nodes", () => {
    const out = [
      ...iterateVisualNodes([node("a"), node("b", "PropertyKey" as NodeKind)]),
    ] satisfies VisualNode[];
    expect(out.map((v) => v.id)).toEqual(["a"]);
  });

  test("recurses into subgraphs", () => {
    const out = [
      ...iterateVisualNodes([
        sg("s", [node("inner"), sg("s2", [node("deep")])]),
        node("top"),
      ]),
    ] satisfies VisualNode[];
    expect(out.map((v) => v.id)).toEqual(["inner", "deep", "top"]);
  });

  test("empty input → empty output", () => {
    expect([...iterateVisualNodes([])]).toEqual([]);
  });
});

File: crates/unsnarl-visual-graph/src/prune/iterate_visual_nodes_test.rs
use super::*;

use crate::prune::test_helpers::{const_binding_node, function_subgraph};
use crate::visual_element_type::NodeTypeTag;
use crate::visual_node::{SyntheticExtras, SyntheticNodeKind, SyntheticVisualNode};

fn non_candidate_node(id: &str) -> VisualNode {
    // SyntheticBeyondDepth is not in ROOT_CANDIDATE_KINDS, so it
    // must not be yielded.
    VisualNode::Synthetic(SyntheticVisualNode {
        r#type: NodeTypeTag::Node,
        id: id.to_string(),
        kind: SyntheticNodeKind::SyntheticBeyondDepth,
        name: id.to_string(),
        line: 1,
        end_line: None,
        is_jsx_element: false,
        unused: false,
        extras: SyntheticExtras::None {},
    })
}

#[test]
fn yields_only_root_candidate_kinds() {
    let elements = [
        VisualElement::Node(const_binding_node("a", "a", 1)),
        VisualElement::Node(non_candidate_node("b")),
    ];
    let out = collect_root_candidates(&elements);
    assert_eq!(out.iter().map(|v| v.id()).collect::<Vec<_>>(), vec!["a"]);
}

#[test]
fn recurses_into_subgraphs() {
    let elements = [
        VisualElement::Subgraph(function_subgraph(
            "s",
            1,
            vec![
                VisualElement::Node(const_binding_node("inner", "inner", 1)),
                VisualElement::Subgraph(function_subgraph(
                    "s2",
                    1,
                    vec![VisualElement::Node(const_binding_node("deep", "deep", 1))],
                )),
            ],
        )),
        VisualElement::Node(const_binding_node("top", "top", 1)),
    ];
    let out = collect_root_candidates(&elements);
    assert_eq!(
        out.iter().map(|v| v.id()).collect::<Vec<_>>(),
        vec!["inner", "deep", "top"]
    );
}

#[test]
fn empty_input_yields_empty_output() {
    let empty: [VisualElement; 0] = [];
    assert!(collect_root_candidates(&empty).is_empty());
}
